### PR TITLE
Support automatic pack namespaces

### PR DESCRIPTION
**Pull request by alexevanczuk**
This is a rework of https://github.com/rubyatscale/stimpack/pull/37 based on feedback from @ngan.

This PR supports automatic pack namespaces! Check out the `README.md` change for more information.

### CODE REVIEW

**Comment by alexevanczuk** [comment]
Hi @fxn! I have some questions for you about some things we are trying to do with zeitwerk. I thought it could be easier to chat over slack, so I've invited you to a shared slack channel in the Gusto org, but we can chat here too.

What we are trying to do in this PR is allow `stimpack`, a gem for allowing libraries of code with fuzzy boundaries (i.e. not gems) to live in `packs/*` (see https://github.com/rubyatscale/stimpack for the file structure), while integrating seamlessly into Rails.

We want to allow `stimpack` users to use the [custom root namespaces](https://github.com/fxn/zeitwerk#custom-root-namespaces) feature of zeitwerk. We are running into two major roadblocks:

# 1) Rails doesn't support configuring custom root namespaces when adding load paths
TLDR: When adding to load paths, [rails calls `push_dir`](https://github.com/rails/rails/blob/v7.0.4/railties/lib/rails/application/finisher.rb#L17-L41) and there is no API available, as far as we can tell, to allow you to add to autoload paths *and* define a custom root namespace at the same time. To get around this, we are basically monkey patching the rails autoloader by defining a `ZeitwerkProxy`, which calls `push_dir` with a given namespace when applicable. We don't believe this is a great solution and would love to be able to do this in a more supported way, either with a proper Rails or a proper Zeitwerk API.

More info here: https://github.com/rubyatscale/stimpack/pull/38/files#diff-fdf4ab10aa75775534c8300ff2e842841eac9949d95b8ed3cf21f3fb5293f86dR10

# 2) Zeitwerk doesn't allow the top-level module to have properly reloaded methods defined on it

At Gusto, we often recommend public APIs in the form of `MyNamespace.my_public_api`. The advantage here is that it's very simple – the top-level namespace holds all/most of the public API. However, [per the note in the docs](https://github.com/fxn/zeitwerk#custom-root-namespaces), "the given root namespace must be non-reloadable."

We tried getting around this by [recommending a user eagerly load a file that reopens the top-level module](https://github.com/rubyatscale/stimpack/pull/38/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R136-R163), but is an unsatisfactory user experience, since besides being awkward, those methods will not be auto-reloaded.

Can you help me understand this technical limitation and if there is any way we can allow zeitwerk to allow defining + auto-reloading top-level modules? I understand that zeitwerk needs to define the module (via implicit namespaces) when a user is using custom root namespaces, but I'm wondering if after defining it, it can see if the user has also defined, and if so, allow it to be auto-reloaded (ideally its definition file path can follow some convention where zeitwerk doesn't assume that file path is pre-namespaces, e.g. instead of `MyPublicApi::MyPublicApi`, it just reads it as `MyPublicApi`).

Thank you!!

**Comment by fxn** [comment]
Hi @alexevanczuk! Thanks very much for the Slack invitation. Since I am normally consulting, a synchronous chat is something that it is difficult to fit in my day. Let's discuss async for now!

Rails API is not on par with Zeitwerk, and that is on purpose. Before, Rails had its own autoloader and it was natural that the framework had config points for it. Now, the autoloader is external and the instances are public interface. So, by now we keep backwards compatibility, but we don't plan to replicate the rest (ignore, collapse, etc.).

Regarding custom namespaces, I plan to something about this for Rails 7.2. Before Rails 7.2, you need to manually remove the directory Rails automatically assumes is an autoload path:

```ruby
# config/initializers/zeitwerk.rb

require 'my_custom_namespace'

path = "#{Rails.root}/app/foo" # has to be a string
ActiveSupport::Dependencies.autoload_paths.delete(path) # won't be needed it 7.2
Rails.autoloaders.main.push_dir(path, namespace: MyCustomNamespace)
```

The reason `MyCustomNamespace` cannot be reloadable is that we need the class or module object before the autoloader is ready to autoload.

As you probably know, Zeitwerk sets `autoload` calls on your behalf. For example, when `setup` works, if it sees `app/models/user.rb`, it does the following:

```ruby
Object.autoload(:User, "#{Rails.root}/app/models/user.rb")
```

In the example setup above, if you have `app/foo/x.rb`, it does

```ruby
MyCustomNamespace.autoload(:X, "#{Rails.root}/app/foo/x.rb"
```

See? We need the module object while we are setting things up, too early to autoload.

**Comment by alexevanczuk** [comment]
Thank you @fxn for your prompt response! 

Your explanation about Rails and Zeitwerk makes sense. It sounds like to support this feature, we basically need to basically put something like `Stimpack.enable_automatic_namespaces!` in `config/initializers/zeitwerk.rb` with the implementation that you described. That actually sounds pretty simple! Does `push_dir` overwrite previous pushes without an issue?

> We need the module object while we are setting things up, too early to autoload.

Yep – that makes sense. Would it be possible that after things are set up, we then go back to those custom root namespaces and autoload them, or some other similar implementation? Is there any conceptual or mechanical reason why we couldn't or shouldn't attempt to do that? I figure you've probably already thought of this, but just trying to see if there is any path forward with this pattern.

Thank you again!

**Comment by fxn** [comment]
@alexevanczuk hey!

> Does push_dir overwrite previous pushes without an issue?

Yes. And the `delete` call in Rails < 7.2 is needed precisely because Rails performs the `push_dir` calls later. If we don't delete, it overrides. In Rails 7.2 I plan to modify Rails initialization so that `push_dir` is invoked for a given autoload path _unless the user already did_ (this is not yet written).

> Would it be possible that after things are set up, we then go back to those custom root namespaces and autoload them, or some other similar implementation?

No. The problem is that you cannot load anything during setup, because things are precisely not yet ready to work. And setting autoloads in the root namespaces is what setting things up is all about. Summary:

1. Zeitwerk needs the root namespaces during setup.
2. You need setup finished in order to autoload code.
3. Therefore, you cannot have a root namespace that has to be autoloaded.

And that is why the interface requires a class or module object, you need to have it already instantiated when `push_dir` is executed.

If you want the namespace to be reloadable, you can do it, but cannot be a custom namespace, needs to live in its own reloadable explicit file.

**Comment by alexevanczuk** [comment]
Awesome, thanks so much for all of the detailed context and explanation @fxn! Really appreciate it.

**Comment by gap777** [comment]
@fxn I tried your recommended approach, and think it's working! @alexevanczuk is going to help me to explore extracting the solution to something we can share with others. Kudos! 🥇 🥳 🎉

**Comment by gap777** [comment]
@fxn Actually.... I have to take back those kudos (temporarily)...  Although the recommended approach works (Zeitwerk can load classes and recognize they exist in namespaces, even though namespace directories are missing in the directory tree), it seems to make every directory non-reloadable. That is, a change in a controller isn't in-use until the running bin/dev server is restarted.

Is this behavior what you expected? 

If so, are there any approaches that would leave the application directories (controller, models, etc) reloadable, even if the module file (eg `packs/inventory/public/inventory.rb`) is not? 

If not, would you be willing to look at my solution and help me understand what I did that made the result deviate from your expectations?

I put this file in `config/initializers/automatic_namespaces.rb`:

```
require 'yaml'

class StimpackExtension
  def enable_automatic_namespaces
    namespaced_packages.each do |package_file|
      pack_root_dir = File.dirname(package_file)
      package_namespace = define_namespace(pack_root_dir)
      pack_directories(pack_root_dir).each do |pack_dir|
        set_namespace_for(pack_dir, package_namespace)
      end
    end
  end

  private

  def set_namespace_for(pack_dir, package_namespace)
    Rails.logger.debug { "Associating #{pack_dir} with namespace #{package_namespace}" }
    ActiveSupport::Dependencies.autoload_paths.delete(pack_dir)
    Rails.autoloaders.main.push_dir(pack_dir, namespace: package_namespace)
  end

  PACK_SUBDIRS = [
    'app/components',
    'app/controllers',
    'app/event_handlers',
    'app/events',
    'app/models',
    'app/public',
    'app/services',
    'app/views'
  ]

  def pack_directories(pack_root_dir)
    PACK_SUBDIRS.map {|subdir| Dir.glob("#{pack_root_dir}/#{subdir}") }
                .flatten
                .select { |pack_dir| File.exist?(pack_dir) }
  end

  def define_namespace(pack_root_dir)
    package_directory = File.basename(pack_root_dir).to_s
    package_name = package_directory.camelize
    Object.const_set(package_name, Module.new)
    package_name.constantize
  end

  def namespaced_packages
    # assuming that packages are nested under organizing directory, like "subsystem" or "component"
    Rails.root.glob('packs/*/*/package.yml').select { |package_file| has_automatic_namespace?(package_file) }
  end

  def has_automatic_namespace?(package_file)
    package_description = YAML.load_file(package_file)
    metadata = package_description["metadata"]
    metadata && metadata["automatic_pack_namespace"]
  end
end

StimpackExtension.new.enable_automatic_namespaces
```

**Comment by fxn** [comment]
Oh, quick followup to clarify.

Zeitwerk does not monitor the file system, the library provides a `reload` method, and the parent application is responsible for invoking it. In this case, the parent application is the Rails framework.

Rails has a file watcher, and from memory there are two types, a simple one that scans, and an evented one. Those watchers get passed directories to watch, and probably nowadays they get what is in `autoload_paths`, but these directories are not there. That's, I think, why you don't see reloads happening.

Later I'll dig into this.

**Comment by fxn** [comment]
Confirmed, those directories are not being watched and therefore Rails does not know it has to reload if you edit a pack.

Fortunately, there is public interface to get reloading working, just execute

```ruby
Rails.application.config.watchable_dirs[pack_dir] = [:rb]
```

in `set_namespace_for`.

In the next version of Rails this won't be needed either, but this will work in both versions.

**Comment by gap777** [comment]
@fxn Brilliant. That worked wonderfully.
